pkg/gadgets/top/block-io/tracer/tracer.go

```diff
@@ -273,32 +281,65 @@ func (t *Tracer) nextStats() ([]*types.Stats, error) {
 	return stats, nil
 }
 
-func (t *Tracer) run() {
+func (t *Tracer) run(ctx context.Context, cancel context.CancelFunc) {
+	count := t.config.Count
 	ticker := time.NewTicker(t.config.Interval)
 
-	go func() {
-	loop:
-		for {
-			select {
-			case <-t.done:
-				break loop
-			case <-ticker.C:
-				stats, err := t.nextStats()
-				if err != nil {
-					t.eventCallback(&top.Event[types.Stats]{
-						Error: err.Error(),
-					})
-					return
-				}
+	for {
+		select {
+		case <-t.done:
+			// TODO: Once we completely move to use Run instead of NewTracer,
+			// we can remove this as nobody will directly call Stop (cleanup).
+			return
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			stats, err := t.nextStats()
+			if err != nil {
+				t.eventCallback(&top.Event[types.Stats]{
+					Error: fmt.Sprintf("getting next stats: %s", err),
+				})
+				return
+			}
 
-				n := len(stats)
-				if n > t.config.MaxRows {
-					n = t.config.MaxRows
+			n := len(stats)
+			if n > t.config.MaxRows {
+				n = t.config.MaxRows
+			}
+			t.eventCallback(&top.Event[types.Stats]{Stats: stats[:n]})
+
+			if count > 0 {
+				count--
+				if count == 0 {
+					if cancel != nil {
+						cancel()
+					}
+					return
 				}
-				t.eventCallback(&top.Event[types.Stats]{Stats: stats[:n]})
 			}
 		}
-	}()
+	}
+}
+
+func (t *Tracer) Run(gadgetCtx gadgets.GadgetContext) error {
+	if err := t.init(gadgetCtx); err != nil {
+		return fmt.Errorf("initializing tracer: %w", err)
+	}
+
+	defer t.cleanup()
+	if err := t.install(); err != nil {
+		return fmt.Errorf("installing tracer: %w", err)
+	}
+
+	// Don't use a context with a timeout but a counter to avoid having to deal
+	// with two timers: one for the timeout and another for the ticker.
+	ctx, cancel := context.WithCancel(gadgetCtx.Context())
+	defer cancel()
+
+	go t.run(ctx, cancel)
+	<-ctx.Done()
```

Comment:
What about something like https://github.com/inspektor-gadget/inspektor-gadget/commit/fd33b0a9b955efbb933595b1491eb4d4a4b7bdcf to avoid creatin gan extra goroutine here that is not needed?

Reply:
Yep, I like it!

cc @flyth

Reply:
Actually, we can avoid creating a new context with the cancel:

```go
	func (t *Tracer) run(ctx context.Context) error {
		// Do not use a context with a timeout but use a counter
		count := t.config.Count
		ticker := time.NewTicker(t.config.Interval)
		defer ticker.Stop()

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				[...]
				# Capture and emit events through the callback
				[...]

				if count > 0 {
					count--
					if count == 0 {
						return nil
					}
				}
			}
		}
	}

	func (t *Tracer) Run(gadgetCtx gadgets.GadgetContext) error {
		if err := t.init(gadgetCtx); err != nil {
			return nil, fmt.Errorf("initializing tracer: %w", err)
		}
		
		defer t.cleanup()
		if err := t.install(); err != nil {
			return nil, fmt.Errorf("installing tracer: %w", err)
		}

		return t.run(gadgetCtx.Context())
	}
```

I updated the PR with this.